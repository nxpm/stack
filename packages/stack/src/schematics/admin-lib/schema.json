{
  "$schema": "http://json-schema.org/draft-07/schema",
  "id": "AdminLib",
  "title": "",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like to use for the library?",
      "pattern": "^[a-zA-Z]{1}.*$"
    },
    "type": {
      "description": "The type of library",
      "type": "string",
      "$default": {
        "$source": "argv",
        "index": 1
      },
      "x-prompt": {
        "message": "Which type of library would you like to create?",
        "type": "list",
        "items": [
          {
            "value": "assets",
            "label": "Assets library"
          },
          {
            "value": "data-access",
            "label": "Data access library"
          },
          {
            "value": "feature",
            "label": "Feature library"
          },
          {
            "value": "style",
            "label": "Style Library"
          },
          {
            "value": "ui",
            "label": "UI Library"
          },
          {
            "value": "util",
            "label": "Utility library"
          }
        ]
      }
    },
    "tags": {
      "type": "string",
      "description": "Add tags to the project (used for linting)",
      "alias": "t"
    },
    "directory": {
      "type": "string",
      "description": "A directory where the project is placed",
      "alias": "d"
    },
    "publishable": {
      "type": "boolean",
      "default": false,
      "description": "Generate a publishable library."
    },
    "buildable": {
      "type": "boolean",
      "default": false,
      "description": "Generate a buildable library."
    },
    "prefix": {
      "type": "string",
      "format": "html-selector",
      "description": "The prefix to apply to generated selectors.",
      "alias": "p"
    }
  },
  "required": ["name"]
}
