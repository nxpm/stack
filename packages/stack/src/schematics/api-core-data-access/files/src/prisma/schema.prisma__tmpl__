datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

generator client {
  provider      = "prisma-client-js"
  binaryTargets = ["native"]
}

model User {
  id                   String           @id @default(cuid())
  createdAt            DateTime         @default(now())
  updatedAt            DateTime         @default(now()) @updatedAt
  role                 Role
  developer            Boolean          @default(false)
  email                String           @unique
  username             String           @unique
  password             String?
  firstName            String?
  lastName             String?
  avatarUrl            String?
  location             String?
  phone                String?
  bio                  String?
}

enum Role {
  Admin
  User
}

