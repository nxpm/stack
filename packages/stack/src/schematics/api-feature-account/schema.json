{
  "$schema": "http://json-schema.org/draft-07/schema",
  "id": "ApiFeatureUser",
  "title": "",
  "type": "object",
  "properties": {
    "appName": {
      "type": "string",
      "description": "Name of the app to connect to",
      "alias": "a"
    },
    "directory": {
      "type": "string",
      "description": "A directory where the project is placed",
      "alias": "d"
    }
  },
  "required": []
}
