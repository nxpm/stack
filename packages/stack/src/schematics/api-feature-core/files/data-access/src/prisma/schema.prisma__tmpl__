datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

generator client {
  provider      = "prisma-client-js"
  binaryTargets = ["native"]
}

model User {
  id        String   @id @default(cuid())
  createdAt DateTime @default(now())
  updatedAt DateTime @default(now()) @updatedAt
  role      Role
  developer Boolean  @default(false)
  username  String   @unique
  password  String?
  firstName String?
  lastName  String?
  avatarUrl String?
  location  String?
  phone     String?
  bio       String?
  emails    Email[]  @relation("Email_belongsTo_Owner")
}

enum Role {
  Admin
  User
}

model Email {
  id            String    @id @default(cuid())
  createdAt     DateTime  @default(now())
  updatedAt     DateTime  @default(now()) @updatedAt
  email         String    @unique
  public        Boolean   @default(false)
  primary       Boolean   @default(false)
  verified      Boolean   @default(false)
  verifyToken   String?
  verifyExpires DateTime?
  owner         User      @relation(name: "Email_belongsTo_Owner", fields: [ownerId], references: [id])
  ownerId       String
  // Enable when this is resolved https://github.com/prisma/prisma/issues/3387
  // @@unique([ownerId, primary])
}
