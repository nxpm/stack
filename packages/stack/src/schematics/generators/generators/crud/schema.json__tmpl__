{
  "cli": "nx",
  "id": "web-module",
  "type": "object",
  "properties": {
    "target": {
      "type": "string",
      "description": "Target folder to create the module",
      "$default": {
        "$source": "argv",
        "index": 0
      }
    },
    "name": {
      "type": "string",
      "description": "Name of the module (in kebab-case)",
      "$default": {
        "$source": "argv",
        "index": 1
      }
    }
  },
  "required": ["target", "name"]
}
