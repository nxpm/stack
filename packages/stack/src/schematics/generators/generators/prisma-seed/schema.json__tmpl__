{
  "cli": "nx",
  "id": "prisma-seed",
  "type": "object",
  "properties": {
    "confirm": {
      "type": "boolean",
      "description": "Library confirm",
      "x-prompt": "Are you sure you want to delete and seed the database?"
    },
    "timeout": {
      "type": "number",
      "description": "Number of seconds to wait before before starting deletion",
      "default": 10
    }
  },
  "required": ["confirm", "timeout"]
}
