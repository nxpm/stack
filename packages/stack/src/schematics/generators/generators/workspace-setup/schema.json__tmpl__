{
  "cli": "nx",
  "id": "workspace-setup",
  "type": "object",
  "properties": {},
  "required": []
}
