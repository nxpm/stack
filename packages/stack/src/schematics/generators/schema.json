{
  "$schema": "http://json-schema.org/draft-07/schema",
  "id": "Init",
  "title": "",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "",
      "default": "web"
    }
  },
  "required": []
}
