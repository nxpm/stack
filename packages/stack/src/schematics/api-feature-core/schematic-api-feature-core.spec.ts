import { Tree } from '@angular-devkit/schematics'
import { SchematicTestRunner } from '@angular-devkit/schematics/testing'
import { createEmptyWorkspace } from '@nrwl/workspace/testing'
import { join } from 'path'

import { ApiFeatureCoreSchematicSchema } from './schema'

describe('api-feature-core schematic', () => {
  let appTree: Tree
  const options: ApiFeatureCoreSchematicSchema = { name: 'test', type: 'feature' }

  const testRunner = new SchematicTestRunner('@nxpm/api-feature-core', join(__dirname, '../../../collection.json'))

  beforeEach(() => {
    appTree = createEmptyWorkspace(Tree.empty())
  })

  it('should run successfully', async () => {
    await expect(
      testRunner.runSchematicAsync('api-feature-core', options, appTree).toPromise(),
    ).resolves.not.toThrowError()
  })
})
