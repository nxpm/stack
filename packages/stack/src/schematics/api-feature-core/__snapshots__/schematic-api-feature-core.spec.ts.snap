// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api-feature-core schematic should run successfully 1`] = `
"import { Module } from '@nestjs/common'

import { TestCoreDataAccessService } from './test-core-data-access.service'

@Module({
  providers: [TestCoreDataAccessService],
  exports: [TestCoreDataAccessService],
})
export class TestCoreDataAccessModule {}
"
`;

exports[`api-feature-core schematic should run successfully 2`] = `
"import { BadRequestException, Injectable, OnModuleDestroy, OnModuleInit } from '@nestjs/common'
import { PrismaClient, Prisma, Role } from '@prisma/client'
import { getGravatarUrl, hashPassword } from '@nxpm/test/core/util'

@Injectable()
export class TestCoreDataAccessService extends PrismaClient implements OnModuleInit, OnModuleDestroy {
  constructor() {
    super()
  }

  async onModuleDestroy() {
    await this.$disconnect()
  }

  async onModuleInit() {
    await this.$connect()
  }

  async createUser(input: Partial<Prisma.UserCreateInput>, emailInput: Partial<Prisma.EmailCreateInput>) {
    const email = emailInput?.email?.trim()
    const existing = await this.findUserByEmail(email)
    if (existing) {
      throw new BadRequestException(\`Can't create user with email \${email}\`)
    }
    const password = hashPassword(input.password)

    // The first user will get the Admin role
    const hasAdmin = await this.user.count({ where: { role: 'Admin' } })
    const role = hasAdmin ? 'User' : 'Admin'

    return this.user.create({
      data: {
        firstName: input.firstName,
        lastName: input.lastName,
        emails: { create: { email, primary: true } },
        username: email,
        avatarUrl: input.avatarUrl || getGravatarUrl(email),
        password,
        role,
      },
    })
  }

  findUserByEmail(email: string) {
    return this.email.findUnique({ where: { email } }).owner({ include: { emails: true } })
  }

  findUserById(userId: string) {
    return this.user.findUnique({ where: { id: userId }, include: { emails: true } })
  }

  findUserByUsername(username: string) {
    return this.user.findUnique({ where: { username }, include: { emails: true } })
  }

  async ensureAdminUser(adminId: string): Promise<boolean> {
    const tenant = await this.findUserById(adminId)
    if (tenant.role !== Role.Admin) {
      throw new Error(\`This operation needs Admin access\`)
    }
    return true
  }

  async usernameDenyList(username: string) {
    const denied = [
      // Add usernames you want to deny
      'admin',
      'administrator',
      'system',
      'owner',
      'info',
      'webmaster',
    ]
    return denied.includes(username)
  }
}
"
`;

exports[`api-feature-core schematic should run successfully 3`] = `
"import { GraphQLIntercomModule } from '@kikstart-playground/graphql-intercom'
import { Module } from '@nestjs/common'
import { GraphQLModule } from '@nestjs/graphql'
import { PubSub } from 'graphql-subscriptions'
import { join } from 'path'
import { TestAccountFeatureModule } from '@nxpm/test/account/feature'
import { TestAuthFeatureModule } from '@nxpm/test/auth/feature'
import { TestUserFeatureModule } from '@nxpm/test/user/feature'

import { TestCoreFeatureController } from './test-core-feature.controller'
import { TestCoreFeatureResolver } from './test-core-feature.resolver'
import { TestCoreFeatureService } from './test-core-feature.service'

@Module({
  imports: [
    GraphQLModule.forRoot({
      autoSchemaFile: join(process.cwd(), 'test-schema.graphql'),
      context: ({ req, res }) => ({ req, res }),
      installSubscriptionHandlers: true,
      playground: {
        settings: {
          'request.credentials': 'include',
        },
      },
      sortSchema: true,
    }),
    GraphQLIntercomModule.forRoot({ pubSub: new PubSub() }),
    TestAccountFeatureModule,
    TestAuthFeatureModule,
    TestUserFeatureModule,
  ],
  controllers: [TestCoreFeatureController],
  providers: [TestCoreFeatureResolver, TestCoreFeatureService],
  exports: [TestCoreFeatureService],
})
export class TestCoreFeatureModule {}
"
`;

exports[`api-feature-core schematic should run successfully 4`] = `
"import { Float, Query, Resolver } from '@nestjs/graphql'
import { TestCoreFeatureService } from './test-core-feature.service'

@Resolver()
export class TestCoreFeatureResolver {
  constructor(private readonly service: TestCoreFeatureService) {}

  @Query(() => Float, { nullable: true })
  uptime() {
    return this.service.uptime()
  }
}
"
`;
